rust/main/hyperlane-base/src/cache/moka/dynamic_expiry.rs

```diff
@@ -0,0 +1,87 @@
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+use chrono::{offset::LocalResult, TimeZone, Utc};
+use moka::Expiry;
+use serde::{Deserialize, Serialize};
+
+/// The type of expiration for a cache entry.
+///
+/// ## Variants
+///
+/// - `Never`: Never expire.
+/// - `AfterDuration`: Expire after a specified duration.
+/// - `AfterTimestamp`: Expire after a specified timestamp.
+/// - `Default`: Use the default expiration. (5 minutes)
+#[allow(missing_docs)]
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
+pub enum ExpirationType {
+    Never,
+    AfterDuration(Duration),
+    AfterTimestamp(u64),
+    Default,
+}
+
+impl From<ExpirationType> for Expiration {
+    fn from(expiration: ExpirationType) -> Self {
+        Expiration {
+            variant: expiration,
+            created_at: Utc::now().timestamp() as u64,
+        }
+    }
+}
+
+/// Expiration to store alongside a cache entry.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Expiration {
+    /// The type of expiration used when the entry was created.
+    pub variant: ExpirationType,
+    /// Unix timestamp when the entry was created.
+    pub created_at: u64,
+}
+
+impl Expiration {
+    /// Get the duration until the entry expires
+    pub fn as_duration(&self) -> Option<Duration> {
+        match self.variant {
+            ExpirationType::AfterDuration(duration) => Some(duration),
+            ExpirationType::AfterTimestamp(timestamp) => {
+                let target_time = UNIX_EPOCH + Duration::from_secs(timestamp);
+                target_time
+                    .duration_since(SystemTime::now())
+                    .ok()
+                    .or(Some(Duration::ZERO))
+            }
+            ExpirationType::Never => None,
+            ExpirationType::Default => Some(Duration::from_secs(60 * 5)),
+        }
+    }
+
+    /// Calculate the time to live for the entry
+    /// Returns None if the entry should never expire or if the expiration time is in the past
```

Comment:
kinda feels like Duration::ZERO should be returned instead of None if the expiration time is in the past?

Reply:
I guess my point is why is `None` used to cover both those very different use cases? How can you disambiguate the two? If we are already past the time where the Expiration should've expired, shouldn't we say that there are 0 more seconds to live?

Iiuc this is just used for a test so it doesn't totally matter though